scalanet/discovery/src/io/iohk/scalanet/discovery/ethereum/v4/DiscoveryNetwork.scala

```diff
@@ -0,0 +1,376 @@
+
+            case UnexpectedError(ex) =>
+              Task.raiseError(new PacketException(ex.getMessage))
+          }
+          .completedL
+      }
+
+      private def handleRequest(
+          handler: DiscoveryRPC[(PublicKey, A)],
+          channel: Channel[A, Packet],
+          remotePublicKey: PublicKey,
+          hash: Hash,
+          payload: Payload.Request
+      ): Task[Unit] = {
+        val caller = (remotePublicKey, channel.to)
+
+        payload match {
+          case Ping(_, _, to, _, maybeRemoteEnrSeq) =>
+            maybeRespond {
+              handler.ping(caller)(maybeRemoteEnrSeq)
+            } { maybeLocalEnrSeq =>
+              channel.send(Pong(to, hash, 0, maybeLocalEnrSeq)).void
+            }
+
+          case FindNode(target, expiration) =>
+            maybeRespond {
+              handler.findNode(caller)(target)
+            } { nodes =>
+              nodes
+                .take(kademliaBucketSize)
+                .grouped(maxNeighborsPerPacket)
+                .toList
+                .traverse { group =>
+                  channel.send(Neighbors(group.toList, 0))
+                }
+                .void
+            }
+
+          case ENRRequest(_) =>
+            maybeRespond {
+              handler.enrRequest(caller)(())
+            } { enr =>
+              channel.send(ENRResponse(hash, enr)).void
+            }
+        }
+      }
+
+      private def maybeRespond[Res](maybeResponse: Task[Option[Res]])(
+          f: Res => Task[Unit]
+      ): Task[Unit] =
+        maybeResponse.attempt.flatMap {
+          case Right(Some(response)) =>
+            f(response)
+          case Right(None) =>
+            Task.unit
+          case Left(NonFatal(ex)) =>
+            // Not responding to this one, but it shouldn't stop handling further requests.
+            Task(logger.error(s"Error handling incoming request: $ex"))
+        }
+
+      private def pack(payload: Payload): Task[Packet] =
+        Packet
+          .pack(payload, privateKey)
+          .fold(
+            err => Task.raiseError(new IllegalArgumentException(s"Could not pack $payload: $err")),
+            packet => Task.pure(packet)
+          )
+
+      private def setExpiration(payload: Payload): Task[Payload] = {
+        payload match {
+          case p: Payload.HasExpiration[_] =>
+            currentTimeMillis.map(t => p.withExpiration(t + expirationMillis))
+          case p =>
+            Task.pure(p)
+        }
+      }
+
+      override val ping = (to: A) =>
+        (localEnrSeq: Option[ENRSeq]) =>
+          peerGroup.client(to).use { channel =>
+            channel
+              .send(
+                Ping(version = 4, from = localNodeAddress, to = toNodeAddress(to), 0, localEnrSeq)
+              )
+              .flatMap { packet =>
+                channel.collectFirstResponse {
+                  case Pong(_, pingHash, _, maybeRemoteEnrSeq) if pingHash == packet.hash =>
```

Comment:
just of caution, did you make sure this wat geth has in pong response ? (and maybe besu:https://github.com/hyperledger/besu)

Reply:
I didn't look at besu, just trinity. It had a special [workaround](https://github.com/ethereum/trinity/blob/dc0ddbb5ee5129b29ea2d4a9444f6a1b60cf09f4/p2p/discovery.py#L1028) for some openethereum 1.10 clients, but I figured since we'll use this in a new testnet in a fork, it won't have to work with old unupgraded clients. geth seems to [calculate](https://github.com/ethereum/go-ethereum/blob/master/p2p/discover/v4wire/v4wire.go#L267) the hash the same way and then [compare](https://github.com/ethereum/go-ethereum/blob/master/p2p/discover/v4_udp.go#L238) to the Pong.

Reply:
we will use this in new test net and on mainnet. So i would check out at least compatibility with geth and besu.(as they make about 90% of etc nodes) openethereum announecd that it removes support for etc so it is probably not that important.

Reply:
I checked besu, the [computation](https://github.com/hyperledger/besu/blob/f9deb54682cb24779bc4919f00b4a953763db8a9/ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/Packet.java#L59-L60) of the packet hash is mostly the same, except for example that they hash the data with keccak256 before signing, where as I imagine [trinity](https://github.com/ethereum/trinity/blob/dc0ddbb5ee5129b29ea2d4a9444f6a1b60cf09f4/p2p/discovery.py#L1323-L1324) takes the hash inside the signing function itself, which is what our `SigAlg.sign` will have to do as well. 

It will also have to make sure that the `Signature` it returns is encoded in the expected way, compatible with [encodeSignature](https://github.com/ethereum/trinity/blob/dc0ddbb5ee5129b29ea2d4a9444f6a1b60cf09f4/p2p/discovery.py#L1323-L1324). I want to use the test vectors I found here and there but at the moment I can only do this in Mantis.

After that the [checking of the pong response](https://github.com/hyperledger/besu/blob/master/ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java#L462) is the same.

Reply:
On a side note: besu doesn't seem to [respect](https://github.com/hyperledger/besu/blob/master/ethereum/p2p/src/main/java/org/hyperledger/besu/ethereum/p2p/discovery/internal/PeerDiscoveryController.java#L555-L556) the packet size limits so we may never accept a neighbors response from them.